contracts/interchain-token-service/src/contract.rs

```diff
@@ -113,16 +127,66 @@ impl InterchainTokenServiceInterface for InterchainTokenService {
             .into()
     }
 
+    fn interchain_token_id(env: &Env, sender: Address, salt: BytesN<32>) -> BytesN<32> {
+        env.crypto()
+            .keccak256(&(PREFIX_INTERCHAIN_TOKEN_ID, sender, salt).to_xdr(env))
+            .into()
+    }
+
     fn deploy_interchain_token(
-        _env: &Env,
-        _caller: Address,
-        _salt: BytesN<32>,
-        _destination_chain: String,
-        _token_metadata: TokenMetadata,
-        _minter: Option<Bytes>,
-        _gas_token: Token,
-    ) -> Result<BytesN<32>, ContractError> {
-        todo!()
+        env: &Env,
+        caller: Address,
+        salt: BytesN<32>,
+        token_meta_data: TokenMetadata,
+        initial_supply: i128,
+        minter: Option<Address>,
+    ) -> Result<Address, ContractError> {
+        caller.require_auth();
+
+        let mut initial_minter: Option<Address> = None;
+
+        if initial_supply > 0 {
+            initial_minter = Some(env.current_contract_address());
+        } else if let Some(ref minter_addr) = minter {
+            ensure!(
+                *minter_addr != env.current_contract_address(),
+                ContractError::InvalidMinter
+            );
+            initial_minter = minter.clone();
+        }
+
+        let deploy_salt = Self::interchain_token_deploy_salt(env, caller.clone(), salt.clone());
+        let token_id = Self::interchain_token_id(env, caller.clone(), salt);
```

Comment:
Note that `interchain_token_id` methods are different between the ITS factory and ITS contracts, the above code is conflating the two

Reply:
I recommend implementing the `interchain_token_id` method to match what ITS factory does, since that's the actual token_id derivation that's relevant here

Reply:
I see. updated.